libs/utils/target_script.py

```diff
@@ -0,0 +1,84 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2015, ARM Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import ast
+import inspect
+
+SCRIPT_NAME = 'remote_script.sh'
+
+class TargetScript(object):
+    """
+    This class provides utility to create and run a script
+    directly on a devlib target.
+    """
+
+    _target_attrs = ['screen_resolution', 'android_id', 'abi', 'os_version', 'model']
+
+    def __init__(self, env):
+        self._env = env
+        self.commands = []
+        
+    # This is made to look like the devlib Target execute()
+    def execute(self, cmd):
+        self.commands.append(cmd)
+
+    # Some commands may require some info about the real target.
+    # For instance, System.{h,v}swipe needs to get the value of
+    # screen_resolution to generate a swipe command at a given
+    # screen coordinate percentage.
+    # Thus, if such a property is called on this object,
+    # it will be fetched from the 'real' target object.
+    def __getattr__(self, name):
+        if name in self._target_attrs:
+            return getattr(self._env.target, name)
+        else:
+            return getattr(super, name)
+            
+    def setup(self):
+        """
+        Push a script to the target
+
+        The script is created and stored on the host,
+        and is then sent to the target.
+
+        :param path: Path where the script will be locally created
+        :type path: str
+        :param actions: List of actions(commands) to run
+        :type actions: list(str)
+        """
+
+        actions = ['set -e'] + self.commands + ['set +e']
+        actions = ['#!{} sh'.format(self._env.target.busybox)] + actions
+        actions = str.join('\n', actions)
+
+        self._remote_path = '{}/{}'.format(self._env.target.executables_directory, SCRIPT_NAME)
+        self._local_path = '{}/{}'.format(self._env.res_dir, SCRIPT_NAME)
+
+        # Create script locally
+        with open(self._local_path, 'w') as script:
+            script.write(actions)
+
+        # Push it on target
+        self._env.target.push(self._local_path, self._remote_path)
+        self._env.target.execute('chmod +x {}'.format(self._remote_path))
+
+    def run(self):
+        """
+        Run the previously pushed script
+        """
+        if self._remote_path:
```

Comment:
Perhaps safer to check with `self._target.file_exists(self._remote_path)` and raise is not?